src/lib/parallel_scan/parallel_scan.mli

```diff
@@ -21,92 +21,202 @@
+          | Empty
+          | Part of 'a (*Only the left component of the job is available yet since we always complete the jobs from left to right*)
+          | Full of
+              { left: 'a
+              ; right: 'a
+              ; seq_no: Sequence_number.Stable.V1.t
+                    (*Update number, for debugging*)
+              ; status: Job_status.Stable.V1.t }
+        [@@deriving sexp, bin_io, version]
       end
+
+      module Latest = V1
+    end
+
+    type 'a t = 'a Stable.Latest.t =
+      | Empty
+      | Part of 'a
+      | Full of
+          { left: 'a
+          ; right: 'a
+          ; seq_no: Sequence_number.Stable.V1.t
+          ; status: Job_status.Stable.V1.t }
+    [@@deriving sexp]
+  end
+
+  module Stable : sig
+    module V1 : sig
+      type 'a t =
+        (Weight.Stable.V1.t * Weight.Stable.V1.t) * 'a Job.Stable.V1.t
+      [@@deriving sexp, bin_io, version]
+    end
+
+    module Latest = V1
+  end
+
+  type 'a t = 'a Stable.Latest.t [@@deriving sexp]
+end
+
+(** An available job is an incomplete job that has enough information for one
+* to process it into a completed job *)
+module Available_job : sig
+  module Stable : sig
+    module V1 : sig
+      type ('a, 'd) t = Base of 'd | Merge of 'a * 'a [@@deriving sexp]
     end
 
+    module Latest = V1
+  end
+
+  type ('a, 'd) t = ('a, 'd) Stable.Latest.t = Base of 'd | Merge of 'a * 'a
+  [@@deriving sexp]
+end
+
+module Space_partition : sig
+  module Stable : sig
+    module V1 : sig
+      type t = {first: int * int; second: (int * int) option}
+      [@@deriving sexp, bin_io, version]
+    end
+
+    module Latest = V1
+  end
+
+  type t = Stable.Latest.t = {first: int * int; second: (int * int) option}
+  [@@deriving sexp]
+end
+
+module Job_view : sig
+  module Node : sig
     module Stable : sig
       module V1 : sig
-        type ('a, 'd) t =
-          | Merge of 'a Merge.Stable.V1.t
-          | Base of ('d * Sequence_no.Stable.V1.t) option
-        [@@deriving bin_io, sexp, version]
+        type 'a t = Base of 'a option | Merge of 'a option * 'a option
+        [@@deriving sexp, bin_io, version]
       end
 
       module Latest = V1
     end
 
-    type ('a, 'd) t = ('a, 'd) Stable.Latest.t =
-      | Merge of 'a Merge.Stable.V1.t
-      | Base of ('d * Sequence_no.Stable.V1.t) option
+    type 'a t = 'a Stable.Latest.t =
+      | Base of 'a option
+      | Merge of 'a option * 'a option
     [@@deriving sexp]
   end
 
-  module Completed_job : sig
-    (** A completed job is either some [Lifted 'a] corresponding to some
-     * [Base 'd] or a [Merged 'a] corresponding to some [Merge ('a * 'a)]
-     *)
-    type 'a t = Lifted of 'a | Merged of 'a [@@deriving bin_io, sexp]
+  module Stable : sig
+    module V1 : sig
+      type 'a t =
+        { position: int
+        ; seq_no: Sequence_number.Stable.V1.t
+        ; status: Job_status.Stable.V1.t
+        ; value: 'a Node.Stable.V1.t }
+      [@@deriving sexp, bin_io, version]
```

Comment:
If there's no registration of the version, you should use `version {unnumbered}` to avoid an unused value warning